File: lib/features/bills/data/bill_repository.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../../core/mock/mock_network.dart';
import '../../saved_billers/data/saved_biller.dart';
import 'fetched_bill.dart';

abstract interface class BillRepository {
  Future<FetchedBill> fetchBill(String billerId, String account);
  Future<List<FetchedBill>> fetchDueBills(List<SavedBiller> saved);
}

class FakeBillRepository implements BillRepository {
  FakeBillRepository(this._network);

  final MockNetwork _network;

  static const _customers = [
    'Ramesh Kumar',
    'Priya Sharma',
    'Amit Patel',
    'Sunita Reddy',
    'Vikram Singh',
  ];

  static int _seedOf(String billerId, String account) =>
      '$billerId|$account'.codeUnits.fold(0, (sum, unit) => sum + unit);

  static FetchedBill _billFor(String billerId, String account) {
    final seed = _seedOf(billerId, account);
    final today = DateTime.now();
    return FetchedBill(
      billerId: billerId,
      account: account,
      customerName: _customers[seed % _customers.length],
      amountPaise: (seed % 4500 + 200) * 100,
      dueDate: DateTime(today.year, today.month, today.day)
          .add(Duration(days: seed % 15 - 3)),
      billNumber: 'BILL-${seed.toRadixString(16).toUpperCase()}-$seed',
    );
  }

  @override
  Future<FetchedBill> fetchBill(String billerId, String account) async {
    await _network.delay();
    return _billFor(billerId, account);
  }

  @override
  Future<List<FetchedBill>> fetchDueBills(List<SavedBiller> saved) async {
    await _network.delay(6);
    return [for (final s in saved) _billFor(s.billerId, s.account)];
  }
}

final billRepositoryProvider = Provider<BillRepository>(
  (ref) => FakeBillRepository(ref.watch(mockNetworkProvider)),
);

File: src/features/bills/data/billRepository.ts
import { MockNetwork } from '@/core/mock/mockNetwork';
import type { Clock } from '@/core/clock/clock';
import type { SavedBiller } from '@/features/saved_billers/data/savedBiller';
import type { FetchedBill } from './fetchedBill';

const customers = ['Ramesh Kumar', 'Priya Sharma', 'Amit Patel', 'Sunita Reddy', 'Vikram Singh'];

const seedOf = (billerId: string, account: string): number =>
  [...`${billerId}|${account}`].reduce((sum, ch) => sum + ch.charCodeAt(0), 0);

export interface BillRepository {
  fetchBill(billerId: string, account: string): Promise<FetchedBill>;
  fetchDueBills(saved: SavedBiller[]): Promise<FetchedBill[]>;
}

export class FakeBillRepository implements BillRepository {
  constructor(private readonly network: MockNetwork, private readonly clock: Clock) {}

  private billFor(billerId: string, account: string): FetchedBill {
    const seed = seedOf(billerId, account);
    const today = this.clock();
    const dueDate = new Date(today.getFullYear(), today.getMonth(), today.getDate() + ((seed % 15) - 3));
    return {
      billerId,
      account,
      customerName: customers[seed % customers.length]!,
      amountPaise: ((seed % 4500) + 200) * 100,
      dueDate,
      billNumber: `BILL-${seed.toString(16).toUpperCase()}-${seed}`,
    };
  }

  async fetchBill(billerId: string, account: string): Promise<FetchedBill> {
    await this.network.delay();
    return this.billFor(billerId, account);
  }

  async fetchDueBills(saved: SavedBiller[]): Promise<FetchedBill[]> {
    await this.network.delay(6);
    return saved.map((s) => this.billFor(s.billerId, s.account));
  }
}
